sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,278 @@
+            },
+        }
+    }
+}
+
+/// Helper to fetch an IMDSv2 session token that is valid for 60 seconds.
+fn fetch_imds_session_token(client: &Client) -> Result<String> {
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
+        .put(uri)
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+    Ok(imds_session_token)
+}
+
+/// Helper to fetch data from IMDS.
+fn fetch_data_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<String> {
+    client
+        .get(uri)
+        .header("X-aws-ec2-metadata-token", session_token)
+        .send()
+        .context(error::ImdsRequest {
+            method: "GET",
+            uri: uri.to_string(),
+        })?
+        .error_for_status()
+        .context(error::ImdsResponse {
+            uri: uri.to_string(),
+        })?
+        .text()
+        .context(error::ImdsText {
+            uri: uri.to_string(),
+        })
+}
+
+/// Returns a list of public keys.
+fn fetch_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    let imds_session_token = fetch_imds_session_token(&client)?;
+
+    info!("Fetching list of available public keys from IMDS");
+    // Returns a list of available public keys as '0=my-public-key'
+    let public_key_list =
+        fetch_data_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to fetch text of available public keys");
+    let public_key_endpoints = build_public_key_endpoints_from_imds(&public_key_list)?;
+
+    info!("Fetching public keys from IMDS");
+    let mut public_keys = Vec::new();
+    for endpoint in public_key_endpoints.iter() {
+        let public_key_text = fetch_data_from_imds(&client, endpoint, &imds_session_token)?;
+        // Simple check to see if the text is probably an ssh key.
+        if &public_key_text[0..3] == "ssh" {
+            debug!("{}", &public_key_text.trim_end());
+            public_keys.push(public_key_text.trim_end().to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls strings for the public keys in IMDS. Since IMDS returns the list
+/// of available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn build_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<String>> {
```

Comment:
nit: we're not talking to IMDS in this function, so perhaps just
```suggestion
fn build_public_key_endpoints(public_key_list: &str) -> Result<Vec<String>> {
```